src/immer.d.ts

```diff
@@ -92,32 +67,32 @@ export interface IProduce {
      * @param {Function} patchListener - optional function that will be called with all the patches produced here
      * @returns {any} a new state, or the initial state if nothing was modified
      */
-    <Base = any, Return = void>(
-        base: Base extends Function ? never : Base,
-        recipe: (this: Draft<Base>, draft: Draft<Base>) => Return,
+
+    /** Curried producer */
+    <
+        Recipe extends (...args: any[]) => any,
+        Params extends any[] = Parameters<Recipe>,
+        T = Params[0],
+    >(
+        recipe: Recipe,
+    ): (state: Immutable<T>, ...rest: Tail<Params>) => Produced<Immutable<T>, ReturnType<Recipe>>
+
+    /** Curried producer with initial state */
+    <
+        Recipe extends (...args: any[]) => any,
+        Params extends any[] = Parameters<Recipe>,
+        T = Params[0],
+    >(
+        recipe: Recipe,
+        initialState: T
+    ): (state?: Immutable<T>, ...rest: Tail<Params>) => Produced<Immutable<T>, ReturnType<Recipe>>
+
+    /** Normal producer */
+    <Base, D = Draft<Base>, Return = void>(
+        base: Base,
+        recipe: (this: D, draft: D) => Return,
         listener?: PatchListener
     ): Produced<Base, Return>
-
-    /** Curried producer with a default value */
-    <Base = any, Rest extends any[] = [], Return = void>(
-        recipe: (this: Base, draft: Base, ...rest: Rest) => Return,
-        defaultBase: Immutable<Base>
-    ): Rest[number][] extends Rest | never[]
-        ? (
-              // The `base` argument is optional when `Rest` is optional.
-              base?: Immutable<Base>,
-              ...rest: Rest
-          ) => Produced<Immutable<Base>, Return>
-        : (
-              // The `base` argument is required when `Rest` is required.
-              base: Immutable<Base> | undefined,
-              ...rest: Rest
```

Comment:
in practice, this `...rest: Rest` always resolved into `...any[]`, as soon as `Base` is set (not inferred) by using `produce<X>((draft, x: number) => {})`